File: code/cvc5mvn/src/main/java/plaid/antlr/TypeVisitor.java
package plaid.antlr;

import plaid.ast.*;

import plaid.PreludeBaseVisitor;
import plaid.PreludeParser.PartyIndexTypeContext;
import plaid.PreludeParser.RecordTypeContext;
import plaid.PreludeParser.StringTypeContext;

import java.util.Map;
import java.util.stream.Collectors;
import java.util.TreeMap;

public class TypeVisitor extends PreludeBaseVisitor<Type> {
    @Override
    public PartyIndexType visitPartyIndexType(PartyIndexTypeContext ctx) {
        return new PartyIndexType();
    }

    @Override
    public StringType visitStringType(StringTypeContext ctx) {
        return new StringType();
    }

    @Override
    public RecordType visitRecordType(RecordTypeContext ctx) {
        Map<Identifier, Type> map = ctx.typedIdent().stream().collect(Collectors.toMap(
                x -> new Identifier(x.ident().getText()),
                x -> visit(x.type())));
        return new RecordType(new TreeMap<>(map));
    }
}

File: code/cvc5mvn/src/main/scala/plaid/prelude/antlr/TypeVisitor.scala
package plaid.prelude.antlr

import plaid.prelude.PreludeBaseVisitor
import plaid.prelude.PreludeParser.*
import plaid.prelude.ast.*

import scala.jdk.CollectionConverters.*

object TypeVisitor extends PreludeBaseVisitor[Type] {

  override def visitPartyIndexType(ctx: PartyIndexTypeContext): PartyIndexType =
    PartyIndexType()

  override def visitStringType(ctx: StringTypeContext): StringType =
    StringType()

  override def visitRecordType(ctx: RecordTypeContext): RecordType =
    RecordType(new java.util.TreeMap(
      ctx.typedIdent().asScala
        .map { x =>
          val id = Identifier(x.ident().getText)
          val tpe = visit(x.`type`())
          id -> tpe
        }
        .toMap.asJava))
}